File: examples/java/transaction/src/main/java/dev/angzarr/examples/transaction/DefaultTransactionLogic.java
package dev.angzarr.examples.transaction;

import com.google.protobuf.Any;
import com.google.protobuf.InvalidProtocolBufferException;
import com.google.protobuf.Timestamp;
import examples.Domains.*;
import dev.angzarr.EventBook;
import dev.angzarr.EventPage;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

import static net.logstash.logback.argument.StructuredArguments.kv;

/**
 * Default implementation of transaction business logic.
 */
public class DefaultTransactionLogic implements TransactionLogic {
    private static final Logger logger = LoggerFactory.getLogger(DefaultTransactionLogic.class);

    @Override
    public TransactionState rebuildState(EventBook eventBook) {
        TransactionState state = TransactionState.empty();

        if (eventBook == null || eventBook.getPagesList().isEmpty()) {
            return state;
        }

        for (EventPage page : eventBook.getPagesList()) {
            if (!page.hasEvent()) {
                continue;
            }
            state = applyEvent(state, page.getEvent());
        }

        return state;
    }

    private TransactionState applyEvent(TransactionState state, Any event) {
        String typeUrl = event.getTypeUrl();

        try {
            if (typeUrl.endsWith("TransactionCreated")) {
                TransactionCreated created = event.unpack(TransactionCreated.class);
                return new TransactionState(
                    created.getCustomerId(),
                    new ArrayList<>(created.getItemsList()),
                    created.getSubtotalCents(),
                    0,
                    "",
                    TransactionState.Status.PENDING
                );
            } else if (typeUrl.endsWith("DiscountApplied")) {
                DiscountApplied applied = event.unpack(DiscountApplied.class);
                return new TransactionState(
                    state.customerId(),
                    state.items(),
                    state.subtotalCents(),
                    applied.getDiscountCents(),
                    applied.getDiscountType(),
                    state.status()
                );
            } else if (typeUrl.endsWith("TransactionCompleted")) {
                return new TransactionState(
                    state.customerId(),
                    state.items(),
                    state.subtotalCents(),
                    state.discountCents(),
                    state.discountType(),
                    TransactionState.Status.COMPLETED
                );
            } else if (typeUrl.endsWith("TransactionCancelled")) {
                return new TransactionState(
                    state.customerId(),
                    state.items(),
                    state.subtotalCents(),
                    state.discountCents(),
                    state.discountType(),
                    TransactionState.Status.CANCELLED
                );
            }
        } catch (InvalidProtocolBufferException e) {
            logger.warn("Failed to unpack event: {}", typeUrl, e);
        }

        return state;
    }

    @Override
    public EventBook handleCreateTransaction(TransactionState state, String customerId, List<LineItem> items)
            throws CommandValidationException {
        if (!state.isNew()) {
            throw CommandValidationException.failedPrecondition("Transaction already exists");
        }

        if (customerId == null || customerId.isBlank()) {
            throw CommandValidationException.invalidArgument("customer_id is required");
        }

        if (items == null || items.isEmpty()) {
            throw CommandValidationException.invalidArgument("at least one item is required");
        }

        int subtotal = items.stream()
            .mapToInt(item -> item.getQuantity() * item.getUnitPriceCents())
            .sum();

        logger.info("creating_transaction",
            kv("customer_id", customerId),
            kv("item_count", items.size()),
            kv("subtotal_cents", subtotal));

        TransactionCreated event = TransactionCreated.newBuilder()
            .setCustomerId(customerId)
            .addAllItems(items)
            .setSubtotalCents(subtotal)
            .setCreatedAt(nowTimestamp())
            .build();

        return createEventBook(event);
    }

    @Override
    public EventBook handleApplyDiscount(TransactionState state, String discountType, int value, String couponCode)
            throws CommandValidationException {
        if (!state.isPending()) {
            throw CommandValidationException.failedPrecondition("Can only apply discount to pending transaction");
        }

        int discountCents = switch (discountType) {
            case "percentage" -> {
                if (value < 0 || value > 100) {
                    throw CommandValidationException.invalidArgument("Percentage must be 0-100");
                }
                yield (state.subtotalCents() * value) / 100;
            }
            case "fixed" -> Math.min(value, state.subtotalCents());
            case "coupon" -> 500; // $5 off
            default -> throw CommandValidationException.invalidArgument("Unknown discount type: " + discountType);
        };

        logger.info("applying_discount",
            kv("discount_type", discountType),
            kv("value", value),
            kv("discount_cents", discountCents));

        DiscountApplied event = DiscountApplied.newBuilder()
            .setDiscountType(discountType)
            .setValue(value)
            .setDiscountCents(discountCents)
            .setCouponCode(couponCode != null ? couponCode : "")
            .build();

        return createEventBook(event);
    }

    @Override
    public EventBook handleCompleteTransaction(TransactionState state, String paymentMethod)
            throws CommandValidationException {
        if (!state.isPending()) {
            throw CommandValidationException.failedPrecondition("Can only complete pending transaction");
        }

        int finalTotal = state.calculateFinalTotal();
        int loyaltyPoints = state.calculateLoyaltyPoints();

        logger.info("completing_transaction",
            kv("final_total_cents", finalTotal),
            kv("payment_method", paymentMethod),
            kv("loyalty_points_earned", loyaltyPoints));

        TransactionCompleted event = TransactionCompleted.newBuilder()
            .setFinalTotalCents(finalTotal)
            .setPaymentMethod(paymentMethod)
            .setLoyaltyPointsEarned(loyaltyPoints)
            .setCompletedAt(nowTimestamp())
            .build();

        return createEventBook(event);
    }

    @Override
    public EventBook handleCancelTransaction(TransactionState state, String reason)
            throws CommandValidationException {
        if (!state.isPending()) {
            throw CommandValidationException.failedPrecondition("Can only cancel pending transaction");
        }

        logger.info("cancelling_transaction", kv("reason", reason));

        TransactionCancelled event = TransactionCancelled.newBuilder()
            .setReason(reason)
            .setCancelledAt(nowTimestamp())
            .build();

        return createEventBook(event);
    }

    private EventBook createEventBook(com.google.protobuf.Message event) {
        EventPage page = EventPage.newBuilder()
            .setNum(0)
            .setEvent(Any.pack(event))
            .setCreatedAt(nowTimestamp())
            .build();

        return EventBook.newBuilder()
            .addPages(page)
            .build();
    }

    private Timestamp nowTimestamp() {
        Instant now = Instant.now();
        return Timestamp.newBuilder()
            .setSeconds(now.getEpochSecond())
            .setNanos(now.getNano())
            .build();
    }
}

File: examples/kotlin/transaction/src/main/kotlin/dev/angzarr/examples/transaction/DefaultTransactionLogic.kt
package dev.angzarr.examples.transaction

import dev.angzarr.EventBook
import dev.angzarr.EventPage
import com.google.protobuf.Any
import com.google.protobuf.Timestamp
import examples.Domains.*
import org.slf4j.LoggerFactory

/**
 * Default implementation of transaction business logic.
 */
class DefaultTransactionLogic : TransactionLogic {

    private val logger = LoggerFactory.getLogger(DefaultTransactionLogic::class.java)

    override fun rebuildState(eventBook: EventBook?): TransactionState {
        if (eventBook == null || eventBook.pagesList.isEmpty()) {
            return TransactionState.empty()
        }

        var state = TransactionState.empty()

        for (page in eventBook.pagesList) {
            val event = page.event ?: continue
            state = applyEvent(state, event)
        }

        return state
    }

    private fun applyEvent(state: TransactionState, event: Any): TransactionState {
        return when {
            event.`is`(TransactionCreated::class.java) -> {
                val e = event.unpack(TransactionCreated::class.java)
                state.copy(
                    customerId = e.customerId,
                    items = e.itemsList,
                    subtotalCents = e.subtotalCents,
                    status = TransactionState.Status.PENDING
                )
            }
            event.`is`(DiscountApplied::class.java) -> {
                val e = event.unpack(DiscountApplied::class.java)
                state.copy(discountCents = e.discountCents, discountType = e.discountType)
            }
            event.`is`(TransactionCompleted::class.java) -> {
                state.copy(status = TransactionState.Status.COMPLETED)
            }
            event.`is`(TransactionCancelled::class.java) -> {
                state.copy(status = TransactionState.Status.CANCELLED)
            }
            else -> state
        }
    }

    override fun handleCreateTransaction(
        state: TransactionState,
        customerId: String,
        items: List<LineItem>
    ): EventBook {
        if (state.exists()) {
            throw CommandValidationException.failedPrecondition("Transaction already exists")
        }

        if (customerId.isEmpty()) {
            throw CommandValidationException.invalidArgument("Customer ID is required")
        }
        if (items.isEmpty()) {
            throw CommandValidationException.invalidArgument("Items are required")
        }

        val subtotal = items.sumOf { it.quantity * it.unitPriceCents }

        logger.info("creating_transaction customer_id={} items={} subtotal={}", customerId, items.size, subtotal)

        val event = TransactionCreated.newBuilder()
            .setCustomerId(customerId)
            .addAllItems(items)
            .setSubtotalCents(subtotal)
            .setCreatedAt(nowTimestamp())
            .build()

        return createEventBook(event)
    }

    override fun handleApplyDiscount(
        state: TransactionState,
        discountType: String,
        value: Int,
        couponCode: String
    ): EventBook {
        if (!state.isPending()) {
            throw CommandValidationException.failedPrecondition("Transaction is not pending")
        }

        val discountCents = if (discountType == "percentage") {
            (state.subtotalCents * value) / 100
        } else {
            value
        }

        logger.info("applying_discount type={} value={} discount_cents={}", discountType, value, discountCents)

        val event = DiscountApplied.newBuilder()
            .setDiscountType(discountType)
            .setValue(value)
            .setDiscountCents(discountCents)
            .setCouponCode(couponCode)
            .build()

        return createEventBook(event)
    }

    override fun handleCompleteTransaction(state: TransactionState, paymentMethod: String): EventBook {
        if (!state.isPending()) {
            throw CommandValidationException.failedPrecondition("Transaction is not pending")
        }

        val finalTotal = state.finalTotal()
        val loyaltyPoints = finalTotal / 100

        logger.info("completing_transaction final_total={} loyalty_points={}", finalTotal, loyaltyPoints)

        val event = TransactionCompleted.newBuilder()
            .setFinalTotalCents(finalTotal)
            .setPaymentMethod(paymentMethod)
            .setLoyaltyPointsEarned(loyaltyPoints)
            .setCompletedAt(nowTimestamp())
            .build()

        return createEventBook(event)
    }

    override fun handleCancelTransaction(state: TransactionState, reason: String): EventBook {
        if (!state.isPending()) {
            throw CommandValidationException.failedPrecondition("Transaction is not pending")
        }

        logger.info("cancelling_transaction reason={}", reason)

        val event = TransactionCancelled.newBuilder()
            .setReason(reason)
            .setCancelledAt(nowTimestamp())
            .build()

        return createEventBook(event)
    }

    private fun createEventBook(event: com.google.protobuf.Message): EventBook {
        val page = EventPage.newBuilder()
            .setNum(0)
            .setEvent(Any.pack(event))
            .setCreatedAt(nowTimestamp())
            .build()

        return EventBook.newBuilder()
            .addPages(page)
            .build()
    }

    private fun nowTimestamp(): Timestamp = Timestamp.newBuilder()
        .setSeconds(System.currentTimeMillis() / 1000)
        .build()
}
